### PR TITLE
Document OOME information

### LINKED ISSUE #3535: Add information about how to mitigate OOME
### Problem description

Memory exhaustion on a host can lead to OOME and killing of important system processes, containers, or the Docker daemon. Document some mitigation and troubleshooting info.

### PR DESCRIPTION
**Pull request by mstanleyjones**
Mitigation and troubleshooting of OOME

Fixes #3535 

cc/ @cpuguy83 @thaJeztah @aaronlehmann @johndmulhausen

### CODE REVIEW

File: engine/admin/index.md [hunk 1]
```diff
@@ -103,7 +112,7 @@ subsystem used:
 ### Enable debugging
 
 There are two ways to enable debugging. The recommended approach is to set the
-`debug` key to `true` in the `daemon.json` file. This method works for every
+`debug` key to `true` in the `daemon.json` file. smethod works for every
```
Comment (cpuguy83) [inline engine/admin/index.md]: smethod?

File: engine/admin/resource_constraints.md [hunk 2]
```diff
@@ -14,6 +14,39 @@ on when you should set such limits and the possible implications of setting them
...
+It is important not to allow a running container to consume too much of the
+host machine's memory. On Linux hosts, if the kernel detects that there is not
+enough memory to perform important system functions, it throws an `OOME`, or
+`Out Of Memory Exception`, and starts killing processes in order to free up
+memory. Any process is subject to killing, including Docker and other important
+applications. This can effectively bring the entire system down if the wrong
+process is killed.
+
+Docker attempts to mitigate these risks by adjusting the OOM priority on the
+Docker daemon so that it will be less likely to be killed than system processes
```
Comment (cpuguy83) [inline engine/admin/resource_constraints.md]: "less likely to be killed than other processes on the system"

File: engine/admin/index.md [hunk 3]
```diff
@@ -84,6 +84,15 @@ non-responsive, you can also
 threads to be added to the daemon log by sending the `SIGUSR` signal to the
 Docker daemon.
 
+### Out Of Memory Exceptions (OOME)
+
+If your containers attempt to use more memmory than the system has available,
+you may experience an Out Of Memory Exception (OOME) and a container, or the
+Docker daemon, might be killed by the kernel OOM killer. To prevent this from
+happening. ensure that your application runs on hosts with adequate memory and
```
Comment (johndmulhausen) [inline engine/admin/index.md]: s/./,

File: engine/admin/index.md (hunk 1 above)
Comment (mstanleyjones) [inline engine/admin/index.md]: WOw how did I do that?

File: engine/admin/index.md [hunk 4]
```diff
@@ -84,6 +84,15 @@ non-responsive, you can also
...
+see
+[Understand the risks of running out of memory](/engine/admin/resource_constraints.md#understand-the-risks-of-running-out-of-memory).
```
Comment (johndmulhausen) [inline engine/admin/index.md]: We should also think about pointing people to the info on resource constraints in the Compose file reference.

File: engine/admin/resource_constraints.md [hunk 5]
```diff
@@ -14,6 +14,42 @@ on when you should set such limits and the possible implications of setting them
...
+The OOM priority on containers is not adjusted. This makes it more likely that
+an individual container will be killed than that the Docker daemon or other
+system processes will be killed.  You should not try to circumvent these
+safeguards by manually setting `--oom-score-adj` to an extreme negative number
+or setting  or `--oom-disable-kill` on a container or the Docker daemon.
```
Comment (thaJeztah) [inline engine/admin/resource_constraints.md]: the docker daemon doesn't have these flags (disable), only oom-weight (don't recall the exact flag name)

**Comment by mstanleyjones** [comment]
Addressed feedback.

File: engine/admin/resource_constraints.md [hunk 6]
```diff
@@ -14,6 +14,42 @@ on when you should set such limits and the possible implications of setting them
...
+- Perform tests to understand the memory requirements of your application before
+  placing it into production.
+- Ensure that your application runs only on hosts with adequate resources.
+- Limit the amount of memory your container can use, as described below
+- Be mindful when configuring swap on your Docker hosts. Swap is slower and
+  less performant than memory but can provide a buffer against running out of
+  system memory. 
+- Consider converting your container to a [service](/engine/swarm/services.md),
+  and using service-level constraints and node labels to ensure that the
+  application runs only on hosts with enough memory
```
Comment (thaJeztah) [inline engine/admin/resource_constraints.md]: great listing :+1:

File: engine/swarm/services.md [hunk 7]
```diff
@@ -522,6 +522,18 @@ the requirement (for instance, if you request 4 CPUs and no node in the swarm
 has 4 CPUs), the service remains in a pending state until a node is available to
 run its tasks.
 
+#### Out Of Memory Exceptions (OOME)
+
+If your service attempts to use more memmory than the swarm node has available,
```
Comment (thaJeztah) [inline engine/swarm/services.md]: typo memmory

File: engine/admin/resource_constraints.md [hunk 8]
```diff
@@ -14,6 +14,43 @@ on when you should set such limits and the possible implications of setting them
...
+host machine's memory. On Linux hosts, if the kernel detects that there is not
+enough memory to perform important system functions, it throws an `OOME`, or
+`Out Of Memory Exception`, and starts killing processes in order to free up
+memory. Any process is subject to killing, including Docker and other important
+applications. This can effectively bring the entire system down if the wrong
+process is killed.
+
+Docker attempts to mitigate these risks by adjusting the OOM priority on the
+Docker daemon so that it will be less likely to be killed than other processes
+on the system The OOM priority on containers is not adjusted. This makes it more
```
Comment (cpuguy83) [inline engine/admin/resource_constraints.md]: Missing a period after system.

**Comment by mstanleyjones** [comment]
@thaJeztah happy?

File: engine/admin/resource_constraints.md [hunk 9]
```diff
@@ -14,6 +14,43 @@ on when you should set such limits and the possible implications of setting them
...
+applications. This can effectively bring the entire system down if the wrong
+process is killed.
+
+Docker attempts to mitigate these risks by adjusting the OOM priority on the
+Docker daemon so that it will be less likely to be killed than other processes
+on the system. The OOM priority on containers is not adjusted. This makes it more
+likely that an individual container will be killed than that the Docker daemon
+or other system processes will be killed.  You should not try to circumvent
+these safeguards by manually setting `--oom-score-adj` to an extreme negative
+number on the daemon or a container, or by setting  or `--oom-disable-kill` on a
```
Comment (thaJeztah) [inline engine/admin/resource_constraints.md]: Missing word here "... or" :sweat_smile:

Comment (thaJeztah) [inline engine/admin/resource_constraints.md]: probably should be removed

Comment (mstanleyjones) [inline engine/admin/resource_constraints.md]: I removed the extra word and squashed the commits.

**Pull request merged**